packages/sdk/src/protocols/base_protocol.ts

```diff
@@ -1,15 +1,206 @@
+          return false;
+        } else {
+          this.log.warn(
+            `Found only ${this.peers.length} peers, expected ${this.numPeersToUse}`
+          );
+          return true;
+        }
+      }
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+    }
+    return true;
+  };
+
+  /**
+   * Starts an interval to maintain the peers list to `numPeersToUse`.
+   * @param interval The interval in milliseconds to maintain the peers.
+   */
+  private async startMaintainPeersInterval(interval: number): Promise<void> {
+    this.log.info("Starting maintain peers interval");
+    try {
+      await this.maintainPeers();
+      this.maintainPeersIntervalId = setInterval(() => {
+        this.maintainPeers().catch((error) => {
+          this.log.error("Error during maintain peers interval:", error);
+        });
+      }, interval);
+      this.log.info(
+        `Maintain peers interval started with interval ${interval}ms`
+      );
+    } catch (error) {
+      this.log.error("Error starting maintain peers interval:", error);
+      throw error;
+    }
+  }
+
+  /**
+   * Maintains the peers list to `numPeersToUse`.
+   */
+  private async maintainPeers(): Promise<boolean> {
+    if (this.maintainPeersLock) {
+      this.log.info("Maintain peers already in progress, skipping");
+      return false;
+    }
+
+    this.maintainPeersLock = true;
+    this.log.info(`Maintaining peers, current count: ${this.peers.length}`);
+    try {
+      const numPeersToAdd = this.numPeersToUse - this.peers.length;
+      if (numPeersToAdd > 0) {
+        await this.findAndAddPeers(numPeersToAdd);
+      }
+      this.log.info(
+        `Peer maintenance completed, current count: ${this.peers.length}`
+      );
+    } catch (error) {
+      this.log.error("Error maintaining peers:", error);
+      throw error;
+    } finally {
+      this.maintainPeersLock = false;
+    }
+    return true;
+  }
+
+  /**
+   * Finds and adds new peers to the peers list.
+   * @param numPeers The number of peers to find and add.
+   */
+  private async findAndAddPeers(numPeers: number = 1): Promise<void> {
+    this.log.info(`Finding and adding ${numPeers} new peers`);
+    try {
+      const additionalPeers = await this.findAdditionalPeers(numPeers);
+      this.peers = [...this.peers, ...additionalPeers];
+      this.log.info(
+        `Added ${additionalPeers.length} new peers, total peers: ${this.peers.length}`
+      );
+    } catch (error) {
+      this.log.error("Error finding and adding new peers:", error);
+      throw error;
+    }
+  }
+
+  /**
+   * Finds additional peers.
+   * Attempts to find peers without using bootstrap peers first,
+   * If no peers are found,
+   * tries with bootstrap peers.
+   * @param numPeers The number of peers to find.
+   */
+  private async findAdditionalPeers(numPeers: number): Promise<Peer[]> {
+    this.log.info(`Finding ${numPeers} additional peers`);
+    try {
+      let newPeers = await this.core.getPeers({
```

Comment:
in general - `getPeers` from `BaseProtocolCore` will get peer only those we are connected to 
is it sufficient? it seems that in some cases we need to initiate new connections, if existing are not enough or not good

Reply:
Actually it seems to be [300](https://github.com/libp2p/js-libp2p/blob/169c9d85e7c9cd65be964b5d08bd618d950f70ee/packages/libp2p/src/connection-manager/constants.ts#L11)

I think we can document it somewhere, at least, in comment section - mentioning that assumption is to have connection manager to populate connected peers.